### PR TITLE
ARTEMIS-2306 Support ActiveMQ5 feature JMSXGroupFirstForConsumer

### PR DESCRIPTION
**Pull request by michaelandrepearce**
Implement using the ActiveMQ5 JMSXGroupFirstForConsumer, property as default, but make it possible for future to make it configurable easily. (Not this PR)
Add test

### CODE REVIEW

**Comment by michaelandrepearce** [comment]
@clebertsuconic small feature, but really useful one from activemq5 with a few people in my org begging for me to implement. If possible to merge before we release again pls. To keep it small so it makes it this week and isn't a big change, i made it use the ActiveMQ5 property name by default.

**Comment by clebertsuconic** [comment]
@michaelandrepearce I think we can still name the next release as 2.7.1, right? this is small feature... I don't see a warrant to name it 2.8.0

**Comment by michaelandrepearce** [comment]
Agreed.

**Comment by clebertsuconic** [comment]
@michaelandrepearce I will wait this before releasing....

**Comment by michaelandrepearce** [comment]
I dont follow sorry. Its ready to merge.

**Comment by clebertsuconic** [comment]
@michaelandrepearce I asked two question, why the name of the property is added as a variable (why not use the constant direct), and move the property definition into MessageUtil alongside with JMSGroupdID definition.

You said agreed, and I assumed you were answering those.

**Comment by michaelandrepearce** [comment]
> I think we can still name the next release as 2.7.1, right? this is small feature... I don't see a warrant to name it 2.8.0

Is all i saw.


Re your comments i made it so we can make in next major release configurable what the property name is. Should note the property in activemq5 isnt actually a jms spec or anything so some people may want to use a different property name.

**Comment by clebertsuconic** [comment]
@michaelandrepearce do you really think we need to add a property for this? 

I would keep it simple and just have it constant, same way with JMSGroupID.

**Comment by michaelandrepearce** [comment]
Yes.

**Comment by clebertsuconic** [comment]
@michaelandrepearce What usecase you have for this? Compatibility with other providers that would apply similar semantics?

Otherwise, I don't see much use for that.

**Comment by michaelandrepearce** [comment]
Because its not jms spec. 

A) what happens if it becomes spec with another name.

B)  non jms use cases where they dont care for jms names but the logical function is useful and in other use case another name is more apt.

**Comment by michaelandrepearce** [comment]
Because its not jms spec. 

A) what happens if it becomes spec with another name.

B)  non jms use cases where they dont care for jms names but the logical function is useful and in other use case another name is more apt.

**Comment by clebertsuconic** [comment]
ok, if you feel strong about eventually making it configurable.. it's fine.

I was trying to avoid leaving dead code and I would only add configuration options when needed.

**Comment by michaelandrepearce** [comment]
I was planning to add in 2.8

**Comment by clebertsuconic** [comment]
@michaelandrepearce if you want to do it already.. just keep a separate commit, and i will branch without the second commit.

**Comment by michaelandrepearce** [comment]
I will do it next week probably. I dont want to hold you up

**Comment by michaelandrepearce** [comment]
@clebertsuconic updated, added the ability to configure also. Have kept 2 commits, as requested, so you have both the defaulted only to cherrypick/branch from, and the fuller more involved configurable part second.

**Comment by clebertsuconic** [comment]
@ig-michaelpearce: just one question I had in a comment. Do you really want to change the wire and API client for that? What about just adding to address settings?

**Comment by michaelandrepearce** [comment]
Again yes, per queue is needed, as like other queue level settings. Where a flow is over multicast address and different consumers are wanting different, e.g. in a migration one consumer subscription (queue) maybe happy with activemq5 header and another consumer subscription (queue) still expecting older other tech's header.

**Comment by michaelandrepearce** [comment]
@clebersuconic - an fyi, i assume your inline comments aren't showing, are you reviewing, but not pressing complete review, or something so theyre not being published, or what ever you're doing theyre not being made public....

**Comment by clebertsuconic** [comment]
I could release with what you have now.. would just need to decide if this warrants a 2.8.0 rename.

File: artemis-core-client/src/main/java/org/apache/activemq/artemis/core/protocol/core/impl/ActiveMQSessionContext.java [hunk 1]
```diff
@@ -876,7 +877,7 @@ public void recreateConsumerOnServer(ClientConsumerInternal consumerInternal,
       // We try to recreate any non-durable or auto-created queues, since they might not be there on failover/reconnect.
       // This allows e.g. JMS non durable subs and temporary queues to continue to be used after failover/reconnection
       if (!queueInfo.isDurable() || queueInfo.isAutoCreated()) {
-         CreateQueueMessage_V2 createQueueRequest = new CreateQueueMessage_V2(queueInfo.getAddress(), queueInfo.getName(), queueInfo.getRoutingType(), queueInfo.getFilterString(), queueInfo.isDurable(), queueInfo.isTemporary(), queueInfo.getMaxConsumers(), queueInfo.isPurgeOnNoConsumers(), queueInfo.isAutoCreated(), false, queueInfo.isExclusive(), queueInfo.isGroupRebalance(), queueInfo.getGroupBuckets(), queueInfo.isLastValue(), queueInfo.getLastValueKey(), queueInfo.isNonDestructive(), queueInfo.getConsumersBeforeDispatch(), queueInfo.getDelayBeforeDispatch(), queueInfo.isAutoDelete(), queueInfo.getAutoDeleteDelay(), queueInfo.getAutoDeleteMessageCount());
```
[the remark this answers is not in the document]
Comment (michaelandrepearce) [reply artemis-core-client/src/main/java/org/apache/activemq/artemis/core/protocol/core/impl/ActiveMQSessionContext.java]: It is settable via address setting which acts as the default. But this allows a per queue should override just like other queue level attributes.

**Comment by michaelandrepearce** [comment]
Upto you. Does it make that much difference?

**Comment by clebertsuconic** [comment]
I will keep it in.. and call it 2.7.1.

**Comment by gemmellr** [comment]
Why not just call it 2.8.0 if there are new features in it, especially ones that seem to change behaviour by default [if using groups]? These dont feel like typical changes for a patch release.

It looks to change what newer Core protocol clients can send on the wire to old servers, and it updates some messages on the server in a way that previously it wouldnt have. The particular message update made being a protocol violation if using AMQP. Can the server side feature be disabled for those that dont want that, or perhaps regardless of protocol run into any unexpected issues with it now being there and looking to disable it?

**Comment by michaelandrepearce** [comment]
Can certainly change it that default to null, and then make it that if null, it doesn't do anything, and only if configured it does.

**Comment by gemmellr** [comment]
Mainly I was just looking for an off-switch, but I do think off-by-default actually makes sense for such a message-manipulating feature, especially given that no existing usage can be using it. That would also nicely address the illegal naming issue, folks pick what to set it to while turning it on.

**Comment by clebertsuconic** [comment]
What about property=“”  meaning off?

Just trying to avoid another new property on confit.

**Comment by michaelandrepearce** [comment]
@clebertsuconic well if we make it off by default , where null/"" then we can make it so people turn it on if by providing the property, then it addresses gemmellr concerns.

**Comment by gemmellr** [comment]
For later reference, the other additions were made via #2625

**Pull request closed without merging**